lld/MachO/ObjC.cpp

```diff
@@ -320,3 +342,1007 @@ void objc::checkCategories() {
+    return false;
+
+  if (!emitAndLinkPointerList(catBodySym, catLayout.classPropsOffset, extInfo,
+                              extInfo.classProps))
+    return false;
+
+  return true;
+}
+
+// This method merges all the categories (sharing a base class) into a single
+// category.
+bool ObjcCategoryMerger::mergeCategoriesIntoSingleCategory(
+    std::vector<InfoInputCategory> &categories) {
+  assert(categories.size() > 1 && "Expected at least 2 categories");
+
+  ClassExtensionInfo extInfo;
+
+  for (auto &catInfo : categories)
+    if (!parseCatInfoToExtInfo(catInfo, extInfo))
+      return false;
+
+  Defined *newCatDef = nullptr;
+  if (!emitCategory(extInfo, newCatDef))
+    return false;
+
+  return true;
+}
+
+bool ObjcCategoryMerger::createSymbolReference(Defined *refFrom, Symbol *refTo,
+                                               uint32_t offset,
+                                               Reloc &relocTemplate) {
+  Reloc r = relocTemplate;
+  r.offset = offset;
+  r.addend = 0;
+  r.referent = refTo;
+  refFrom->isec->relocs.push_back(r);
+
+  return true;
+}
+
+bool ObjcCategoryMerger::collectAndValidateCategoriesData() {
+  for (InputSection *sec : allInputSections) {
+    if (sec->getName() != section_names::objcCatList)
+      continue;
+    ConcatInputSection *catListCisec = dyn_cast<ConcatInputSection>(sec);
+    if (!catListCisec)
+      return registerError(
+          "__objc_catList InputSection is not a ConcatInputSection");
+
+    for (const Reloc &r : catListCisec->relocs) {
+      auto *sym = cast<Defined>(r.referent.get<Symbol *>());
+      if (!sym || !sym->getName().starts_with("__OBJC_$_CATEGORY_"))
+        continue; // Only support ObjC categories (no swift + @objc)
+
+      auto *catBodyIsec =
+          dyn_cast<ConcatInputSection>(r.getReferentInputSection());
+      if (!catBodyIsec)
+        return registerError(
+            "Category data section is not an ConcatInputSection");
+
+      if (catBodyIsec->getSize() != catLayout.totalSize) {
+        std::string err;
+        llvm::raw_string_ostream OS(err);
+        OS << "Invalid input category size encountered, category merging only "
+              "supports "
+           << catLayout.totalSize << " bytes";
+        OS.flush();
+        return registerError(err.c_str());
+      }
+
+      // Check that the category has a reloc at 'klassOffset' (which is
+      // a pointer to the class symbol)
+
+      auto *classReloc = catBodyIsec->getRelocAt(catLayout.klassOffset);
+      if (!classReloc)
+        return registerError("Category does not have a reloc at klassOffset");
+
+      auto *classSym = classReloc->referent.get<Symbol *>();
+      InfoInputCategory catInputInfo{catBodyIsec, catListCisec, r.offset};
+      categoryMap[classSym].push_back(catInputInfo);
+
+      if (!collectCategoryWriterInfoFromCategory(catInputInfo))
+        return false;
+    }
+  }
+
+  for (auto &entry : categoryMap) {
+    if (entry.second.size() > 1) {
```

Comment:
Can we simply delete a single entry case instead of keeping them in `categoryMap` and ignoring it?

Reply:
> I was also thinking if in future we want to implement the "merge into base class directly" then we would need this style of code since we'll also want to preserve 1-size buckets.

Okay. Let's keep it.